### PR TITLE
Custom and remote emojis 

### PR DESCRIPTION
**Pull request by Zeerooth**
It's not ready to be merged or even reviewed, but I'd like to already open this draft PR to keep track of the progress and address potential feedback. 

How it works:
- Server saves all the emojis that it gets through through federated posts (to confirm that they are genuine they get resolved as an AP type, just like accounts are)
- Users then are able to use :emoji@example.com: syntax to insert the emojis into posts (unfortunately it doesn't work for emojis that the server doesn't know about yet as there's no webfinger like for the accounts)
- These emojis get translated into :emoji__example_com: (neither mastodon nor misskey allow special characters like @ or . which sucks) and inserted into the tag list, with id and image url pointing to the original resource.
- Voila, it works

Things that are already implemented and work:
- [x] Parser for emojis in posts
- [x] DB tables 
- [x] Barebones emoji service

Things that probably work but need to be tested:
- [x] Emoji federation - mastodon actually exposes emojis as a resolvable AP type. I implemented a basic resolver ~~but haven't tested it yet. Also who knows how it behaves with misskey or pleroma for example.~~ it works!

To be added:
- [x] GET /api/v1/custom_emojis
- [x] GET /api/v1/statuses/{id} <- needs the emoji list
- [x] Update emojis during post updates 
- [x] More tests
- [x] Expose the emojis as a resolvable AP type

Out of scope for now (I'll make an issue for it later)
- Resolve emojis in account properties, such as description, display name, fields
- Make it easy to add local emojis through the CLI and/or admin panel (if we ever end up making one)

Issue ref: https://github.com/kitsune-soc/kitsune/issues/105

### CODE REVIEW

**Comment by aumetra** [comment]
It indeed does work, heh

**Comment by Zeerooth** [comment]
hmmm I have no idea why the tests keep timing out

**Comment by Zeerooth** [comment]
@aumetra could you take a look at this PR when you've got some time pls?

**Comment by aumetra** [comment]
> could you take a look at this PR when you've got some time pls?

Will do on my commute home

File: crates/kitsune-core/src/activitypub/mod.rs
Comment (aumetra) [inline crates/kitsune-core/src/activitypub/mod.rs:95]: This `.zip` here doesn't seem quite correct. If emoji were filtered in the `.filter_map` step above, there would be a mismatch between the left and right values.

Consider passing them down inside a tuple from the future.

File: crates/kitsune-core/src/mapping/activitypub/object.rs
Comment (aumetra) [inline crates/kitsune-core/src/mapping/activitypub/object.rs:278]: Small sidenote: Do we need that type annotation here?

File: crates/kitsune-core/src/mapping/activitypub/object.rs
Comment (aumetra) [inline crates/kitsune-core/src/mapping/activitypub/object.rs:291]: Async block indirection. We don't need that here

File: crates/kitsune-core/src/mapping/mastodon/sealed.rs
Comment (aumetra) [inline crates/kitsune-core/src/mapping/mastodon/sealed.rs:625]: Consider destructuring `self` inside the block for more readable code.

I.e `let (emoji, attachment, timestamp) = self;`

File: crates/kitsune-core/src/service/custom_emoji.rs
Comment (aumetra) [inline crates/kitsune-core/src/service/custom_emoji.rs:141]: Any reason we box here?

File: lib/post-process/src/lib.rs
Comment (aumetra) [inline lib/post-process/src/lib.rs:46]: Actually we don't need to enforce prefixes and postfixes here, since custom emojis are valid, even when not separated properly from other post components

**Review by aumetra — COMMENTED** [review]
Some initial comments. I'll look at it closer later

File: crates/kitsune-core/src/activitypub/mod.rs
Comment (aumetra) [reply crates/kitsune-core/src/activitypub/mod.rs:95]: That should also make it possible to get rid of the intermediate collect

**Comment by Zeerooth** [comment]
@aumetra thanks so much for these comments, I addressed everything now :)

BTW there is one more aspect of this PR that I was thinking about - for the custom emoji table I decided to not save the domain for local emojis (it's NULL), but I saw that in other tables, for example for user accounts, the domain part is not nullable and saved even for local accounts. Should I change it accordingly for custom emojis?

**Comment by aumetra** [comment]
> Should I change it accordingly for custom emojis?

I don't think that's necessary here. We store the domains on accounts and users because of potential multi-tendant support (i.e. one instance with multiple domains).

But not everyone will be allowed to add custom emoji anyway, just the moderators and admins of the overall system, so might as well not store the domain here.

**Comment by Zeerooth** [comment]
Right, then I'll leave it as it is

**Comment by aumetra** [comment]
@Zeerooth would you mind rebasing this? I don't really have any more comments to this

**Review by aumetra — APPROVED** [review]

**Comment by aumetra** [comment]
Merged main into it. Do you have any last things you wanna add before I hit the merge button?

**Comment by Zeerooth** [comment]
> Merged main into it. Do you have any last things you wanna add before I hit the merge button?

Not really, just that you're too quick, I merged it myself but then I try to push and see that you already pushed to this branch lol

Thanks though, you can merge it now :)

**Pull request merged**
